File: C/Problem0014.c
/**
 * Problem0014 - Longest Collatz sequence
 * 
 * https://github.com/agileshaw/Project-Euler
 **/

#include <stdio.h>

unsigned long long collatz(unsigned long long num)
{
    unsigned long long count = 0;

    while (num != 1) {
        if (num % 2 == 0)
            num /= 2;
        else
            num = 3 * num + 1;
        count++;
    }

    return count;
}

int main()
{
    unsigned long long i, count, max = 0, result;

    for (i = 2; i < 1000000; i++) {
        count = collatz(i);
        if (count > max) {
            max = count;
            result = i;
        }
    }

    printf("The number which generates the longest collatz sequence is: %llu\n", result);

    return 0;
}

File: Perl/Problem0014.pl
# Problem0014 - Longest Collatz sequence
#
# https://github.com/agileshaw/Project-Euler

use strict;
use warnings;

my $result;
my $count;
my $max = 0;
sub collatz;

for my $i (2..1000000) {
    $count = collatz($i);
    if ($count > $max) {
        $max = $count;
        $result = $i;
    }
}

print "The number which generates the longest collatz sequence is: $result\n";

sub collatz
{
    my $num = shift;
    my $tmp = $num;
    my $count = 0;

    while ($tmp != 1) {
        if ($tmp % 2 == 0) {
            $tmp /= 2;
        } else {
            $tmp = 3 * $tmp + 1;
        }
        $count++;
    }

    return $count;
}